### ISSUE TITLE
napari exit callback

### ISSUE THREAD

**Issue by JosephBendahan** [napari/napari#5687]
Hi,
I would like to perform some cleanup before closing a napari application.
What is the name of the exit callback function that would allow doing this when window is closed in any of the possible ways (Exit, clicking the x)?
Thank you.

**Comment by brisvag** [comment]
Hi @JosephBendahan, I don't think we have a public way to do that. You probably need to connect to raw qt `closing` events somwhere on the qt viewer :/

**Comment by Czaki** [comment]
There is such event in QApplication

https://doc.qt.io/qt-6/qguiapplication.html#lastWindowClosed

**Comment by JosephBendahan** [comment]
Thank you very much for finding the function!
Could you please tell us how to access this function within Python?
In other words, what package needs to be imported and do we include the code under "def lastWindowClosed():?

Joseph

**Comment by Czaki** [comment]
I hope that this short example will be enough. But If not, then feel free to ask a question. 

```python
from napari import Viewer, run
from qtpy.QtWidgets import QApplication

def on_close_callback():
    print('Viewer closed')

viewer = Viewer()
app = QApplication.instance()

app.lastWindowClosed.connect(on_close_callback)  # this line is connection to signal

run()
```

Please note that the connection needs to be done after `QApplication` object is created. So performing this during the import of the library may not be the best idea. A good place could be the first creation of your widget or something similar when you are sure that the napari Viewer object already exists.

**Comment by JosephBendahan** [comment]
It works great.
Thank you very much!

**Comment by jni** [comment]
If I'm not mistaken, this wouldn't work in IPython when you have both matplotlib and napari windows open, is that correct @Czaki? ie I think we should add a viewer.closed event anyway?

**Comment by Czaki** [comment]
@jni You are right. But In the scenario where you have expensive shared resources (ex heavy DL model) then you do not want to clean on napari windows close, but on close of last napari window. 

Otherwise, why not clean on close of the widget?

**Comment by brisvag** [comment]
> think we should add a viewer.closed event anyway?

Yup, there's at least 2 old issues I found yesterday abot this :P (and at least 1 of mine that touched this tangentially that I could not find)

**Comment by Czaki** [comment]
But could you explain when it is needed and cannot depend on the deletion of the widget? or delete of last Qt object (or link proper issue)?

**Comment by brisvag** [comment]
some related issues I could find:
- #4623
- #2110
- #3184

**Comment by jni** [comment]
> Otherwise, why not clean on close of the widget?

I'm not sure I'm following your argument @Czaki but mainly I think that we should provide a familiar API for users and allow them to capture a "done with this viewer" signal that does not require delving into our Qt classes. This is even more important if we eventually move to supporting multiple front-ends.

**Comment by Czaki** [comment]
The scenario from #4623 looks like one that would like to receive an event when the last viewer is closed. 

My main problem here is that the longer I think about this callback, the worse the idea. 

Either some resource is connected to a given widget and should be cleaned on the close of the device or belongs to the session and should be cleaned at the end of the session. 

The simple use case is a notebook when the user spawns multiple viewers during the workflow. And bad cleaning after close first notebook may lead to multiple problems. 

Maybe we should suggest overload `__del__` method for perform clean and if something is viewer specific then store it in `WeakKeyDict[Viewer, object]`

**Comment by jni** [comment]
I don't think "cleaning up" is the only use case here @Czaki. I might want to make a beep using [`chime`](https://pypi.org/project/chime/#theming). Or I might want to send someone an email with the name of all the layers. Or, knowing that the number of points in a particular layer can no longer be changed (because the viewer is closed), I can trigger some computation based on those points. etc etc. Of course, all these things can be achieved with a button, but some users might consider it more ergonomic to not have to open the right widget panel for anything.

**Comment by Czaki** [comment]
Ok. I will not opose of adding such event. 

I'm still not sure if it is the best solution. 

So the event should be triggered before any cleaning happen? But in a way that will not allow preventing closing?

**Comment by jni** [comment]
> But in a way that will not allow preventing closing?

ha, nice call... Indeed if the callback has an error I don't know what happens here 😬. I suppose this needs to be a special event subclass (RobustEmitter?) that calls callbacks in a try/finally block.

**Comment by jacopoabramo** [comment]
Greetings, I could also make use of a callback triggered when the viewer is closed: my plugin contains handles to external hardware objects (cameras), and I noticed that when closing down napari from the user interface the handles are not released. Is there any way within a plugin to notify of the close event? Could it be possible to recover the QApplication event and re-use it?

I usually open napari from command line, so to properly close it I still have to shut down the terminal entirely.

**Comment by Czaki** [comment]
Did the solution form this post does not work in your case: https://github.com/napari/napari/issues/5687#issuecomment-1507606948
?

**Comment by jacopoabramo** [comment]
> Did the solution form this post does not work in your case: [#5687 (comment)](https://github.com/napari/napari/issues/5687#issuecomment-1507606948) ?

No, because for how I developed the plugin everything is enclosed within it, and I don't want to control the close event with external code but rather from the UI event of actually closing the window

**Comment by Czaki** [comment]
This is directly event of the close of the last window. 
And if we expose such an event through napari API it will be a simple wrapper. 

As you wrote about releasing the device (camera) then, you may want to wait with the release until the close last window. 

Or in your case, the camera is glued to a given viewer?

**Comment by jacopoabramo** [comment]
I apologize, maybe I'm getting confused. Would that mean that I can get the current instance of the QApplication in which the viewer lives from the plugin itself?

i.e.

```python
from napari import Viewer
from qtpy.QtWidgets import QApplication, QWidget

class MyWidget(QWidget):
    def __init__(self, viewer: Viewer) -> None:
        super().__init__()
        self.viewer = viewer
        self.appInstance = QApplication.instance()
        self.appInstance.lastWindowClosed.connect(self.on_close_callback)  # this line is connection to signal

    def on_close_callback(self) -> None:
        print('See ya chums')
```

**Comment by jacopoabramo** [comment]
> I apologize, maybe I'm getting confused. Would that mean that I can get the current instance of the QApplication in which the viewer lives from the plugin itself?
> 
> i.e.
> 
> ```python
> from napari import Viewer
> from qtpy.QtWidgets import QApplication, QWidget
> 
> class MyWidget(QWidget):
>     def __init__(self, viewer: Viewer) -> None:
>         super().__init__()
>         self.viewer = viewer
>         self.appInstance = QApplication.instance()
>         self.appInstance.lastWindowClosed.connect(self.on_close_callback)  # this line is connection to signal
> 
>     def on_close_callback(self) -> None:
>         print('See ya chums')
> ```

I just tested this out. After launching and trying to open my plugin from napari (because I run it from the UI), napari just crashes.

**Comment by Czaki** [comment]
> I just tested this out. After launching and trying to open my plugin from napari (because I run it from the UI), napari just crashes.

eeeee

The code looks correct. I do not see how it could crash. I'm open even on even some short calls to try to help. 

Could you paste logs?

**Comment by jacopoabramo** [comment]
Ok let me explain and give some context, my development process is a bit unhortodox so I may be doing something wrong.

The plugin in question is [this one](https://github.com/jacopoabramo/napari-live-recording). I usually do my development, then simply install the plugin manually via `pip install`. I then open napari and load the plugin via `Plugins > Live recording` (the name under which the plugin is registered via npe2).

This is the entry point with the modifications I just tested:

```python
from qtpy.QtWidgets import QWidget, QApplication
from napari.viewer import Viewer
from napari_live_recording.ui import ViewerAnchor
from napari_live_recording.control import MainController

class NapariLiveRecording(QWidget):
    def __init__(self, napari_viewer: Viewer) -> None:
        super().__init__()
        self.app = QApplication.instance()
        self.mainController = MainController()
        self.anchor = ViewerAnchor(napari_viewer, self.mainController)
        self.setLayout(self.anchor.mainLayout)
        self.app.aboutToQuit.connect(self.on_close_callback)

    def on_close_callback(self) -> None:
        for cameraKey in self.mainController.deviceControllers.keys():
            self.mainController.deleteCamera(cameraKey)
```

I would post the traceback but none is produced, so I'm puzzled by this - unless you're referring to other logs?

Everything just closes up once I try to show the plugin. The cameras are not glued to the viewers anyway, there are some objects acting as middleman to pass the acquired data to the viewer.

Plus, I could not find the lastWindowClosed signal, so I instead tried the aboutToQuit signal. Same result. I'll go with the oldstyle print to see if ti crashes due to the fetching of the QApplication instance.

**Comment by Czaki** [comment]
Something strange. I just cloned your project. Fixed bug in `napari-live-recording/src/napari_live_recording/control/devices/__init__.py:` (lack of converting PosixPath on string caused plugin crashing on my machine. Replace code from the repository with code from your post and everything works. 

Which version of Qt did you have installed?

**Comment by jacopoabramo** [comment]
> Something strange. I just cloned your project. Fixed bug in `napari-live-recording/src/napari_live_recording/control/devices/__init__.py:` (lack of converting PosixPath on string caused plugin crashing on my machine. Replace code from the repository with code from your post and everything works.

Could you point me to the exact line? Or alternatively open an issue?
 
> Which version of Qt did you have installed?

Name: PyQt5
Version: 5.15.7

**Comment by jacopoabramo** [comment]
Ok I think this has nothing to do with the crashing. I had some extra changes on the codebase and this was still causing issues. I'll try again when I stabilized the situation.

**Comment by Czaki** [comment]
> Could you point me to the exact line? Or alternatively open an issue?

I forget to copy `13`

Here is issue https://github.com/jacopoabramo/napari-live-recording/issues/17

**Comment by jacopoabramo** [comment]
@Czaki ok, now it doesn't crash (the error was with a wrongly implemented combo box). Still though the callback seems to be not triggered at all. I'll try again with the `lastWindowClosed` signal as you initially suggested

EDIT: yep! That did the trick! Apparently `aboutToQuit` does not work as I expected.

**Comment by Czaki** [comment]
I also was not able to make `aboutToQuit` work.
